[Security] Add the ability for voter to return decision reason

yellow1912 opened the pull request:
Sometimes we need to know why the voter makes a certain decision.
This changes allows us to return a vote object which contains a reason.

| Q             | A
| ------------- | ---
| Branch?       | 5.4 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | Fix #27995, #26343, #35592
| License       | MIT
| Doc PR        | n/a

This is a PR continuing the work started by @maidmaid and @noniagriconomie in #35592

All the code is contributed by azjezz, I only merge the changes into 5.4 branch since he/she does not have time right now.
(Original PR: https://github.com/symfony/symfony/pull/40711)

The initial PR is meant to see if all tests passed (I expect tests to fail). More changes will come later since this PR contains BC.

yellow1912 committed:
Add the ability for voter to return decision reason

Sometimes we need to know why the voter makes a certain decision.
This changes allows us to return a vote object which contains a reason.

yellow1912 committed:
Create mock with DecisionManager class

We cannot use the interface at the moment because the new method
is not available in the interface yet.

We also cannot put the new method inside the interface because it
can break current/old code.

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -75,7 +108,7 @@ abstract protected function supports(string $attribute, $subject);
      * @param mixed $subject
      *
-     * @return bool
+     * @return Vote Returning a boolean is deprecated since Symfony 5.1. Return a Vote object instead.
```
Clearly not 5.1

yellow1912 replied:
Changing to 5.4 now

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -48,24 +49,56 @@ public function vote(TokenInterface $token, $subject, array $attributes)
+    /**
+     * Creates a granted vote.
+     */
+    public function grant(string $reason = '', array $parameters = []): Vote
```
Why are those public methods ? They look like helpers for the child class, and so should be protected

yellow1912 replied:
I will make the protected.

stof commented on src/Symfony/Component/Security/Core/Authorization/AccessDecisionManagerInterface.php:
```diff
@@ -17,16 +17,20 @@
+     * @deprecated since 5.3, use {@see getDecision()} instead.
      */
-    public function decide(TokenInterface $token, array $attributes, $object = null);
+    public function decide(TokenInterface $token, array $attributes, $object = null): bool;
```
Adding the return type in an interface is a BC break. That's a no-go (especially to deprecate it)

stof commented on src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php:
```diff
@@ -65,7 +65,7 @@ public function __construct(TokenStorageInterface $tokenStorage, /*AccessDecisio
      * @throws AuthenticationCredentialsNotFoundException when the token storage has no authentication token and $exceptionOnNoToken is set to true
      */
-    final public function isGranted($attribute, $subject = null): bool
+    final public function isGranted($attribute, $subject = null): AccessDecision
```
This is a BC break

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/AccessTrait.php:
```diff
@@ -0,0 +1,41 @@
+trait AccessTrait
+{
+    /** @var int One of the VoterInterface::ACCESS_* constants */
+    protected $access;
```
why protected ? This does not make sense as it forces classes using that trait to expose that field to their child classes (while we even have a getter)

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/AccessTrait.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * @author Dany Maillard <danymaillard93b@gmail.com>
+ */
+trait AccessTrait
```
What is the goal of that trait ? And should it really be part of the public API of the component ?

yellow1912 replied:
huhm, I'm not sure (since this is really not my code. I only merge it to 5.4. I see it is being used in 2 other files: AccessDecision and Vote. Should I simply copy the code over to those 2 files? Should be quick.

stof replied:
To me, yes, you should copy it. The code is simple getter, so there is no harm duplicating it. And it avoids having to maintain BC on that trait.

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/Vote.php:
```diff
@@ -0,0 +1,71 @@
+        $this->parameters = $parameters;
+    }
+
+    public static function create(int $access, string $reason = '', array $parameters = []): self
```
What is the benefit of this compared to the constructor ?

yellow1912 replied:
I'm not sure. The original contributor is using private for __construct. I'm unsure why he/she made that decision though. I can remove this method then update code everywhere? My guess is that he/she wants all methods in Vote to be static.

stof replied:
> My guess is that he/she wants all methods in Vote to be static.

And that's quite the opposite of existing Symfony APIs. In some version (I don't remember which one, check the changelogof HttpFoundation if you care about it), we deprecated `Response::create()` and similar methods that were duplicating the constructor signature.

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -48,24 +49,56 @@ public function vote(TokenInterface $token, $subject, array $attributes)
      * @param mixed  $subject   The subject to secure, e.g. an object the user wants to access or any other PHP type
      *
-     * @return bool
+     * @return bool True if the attribute and subject are supported, false otherwise
```
this does not provide benefits. The description of the function already explains what `supported` means, and I guess nobody will have the idea that we might return `false` as the boolean representing supported things.

stof commented on src/Symfony/Component/Security/Core/Event/VoteEvent.php:
```diff
@@ -51,7 +61,7 @@ public function getAttributes(): array
     }
 
-    public function getVote(): int
+    public function getVote(): Vote
```
This is a BC break for event listeners calling this getter.

yellow1912 replied:
I wonder how this should be resolved. Perhaps I can add a temporary method to return the vote object, and for getVote I will just return int to avoid breaking things?

yellow1912 replied:
I reverted to int, I added a temporary method to get the vote object. Hope it works for now.

stof commented on src/Symfony/Component/Security/Core/Exception/AccessDeniedException.php:
```diff
@@ -57,4 +62,26 @@ public function setSubject($subject)
+            return $vote->getReason();
+        }, $accessDecision->getDeniedVotes());
+
+        $this->message = 'Access Denied: '.rtrim(' '.implode(' ', $reasons));
```
What if a custom message was passed in the constructor ?

yellow1912 replied:
Can you please clarify this comment so I can fix?

stof replied:
see the constructor arguments. We can customize the message to use something else than `Access Denied`, but this is lost if this setter is called.

yellow1912 committed:
Fix phpunit test errors

1. Add missing getDecision method to AuthorizationChecker
2. Do not force int return on TraceableVoter::vote for now

yellow1912 committed:
Standardize Voter code

1. Change internal methods from public to protected
2. Remove useless phpdoc

yellow1912 committed:
Remove AccessTrait

yellow1912 committed:
Change $access property to private as requested

yellow1912 committed:
Remove bool return to avoid BC

yellow1912 committed:
Fix return typehint for decide method

yellow1912 committed:
Switch to using constructor for Vote

yellow1912 committed:
Fix AccessDeniedException

Add missing message passed through constructor

yellow1912 committed:
Fix coding standards

yellow1912 committed:
Fix test results

yellow1912 committed:
Fix deprecation to match with test

yellow1912 committed:
Fix decide method deprecation test matching

noniagriconomie commented:
@yellow1912 thank you taking over this!
As I mentionned on the other PR, my "work" here is only PR review :)

Feel free to ping me when you consider this PR finaly reviewable, I will be glad to do so 👍

yellow1912 committed:
Fix tests in AccessListenerTest

yellow1912 committed:
Add missing AbstractToken in AccessListenerTest

yellow1912 committed:
getDecisionLog return object, not boolean

yellow1912 committed:
Attempt to fix test in AccessListenerTest

yellow1912 committed:
Fix testAccessDecisionManagerCalledByVoter

yellow1912 committed:
Add mising class in test

yellow1912 committed:
Fix coding standards

yellow1912 committed:
Fix issue with final class mock

yellow1912 committed:
Fix issue with final class mock

yellow1912 committed:
Fix issue with final class mock

yellow1912 committed:
Update version in deprecation message

yellow1912 committed:
Fix BC with VoteEvent

noniagriconomie commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -48,24 +49,54 @@ public function vote(TokenInterface $token, $subject, array $attributes)
-            if ($this->voteOnAttribute($attribute, $subject, $token)) {
+            $decision = $this->voteOnAttribute($attribute, $subject, $token);
+            if (\is_bool($decision)) {
+                trigger_deprecation('symfony/security-core', '5.3', 'Returning a boolean in "%s::voteOnAttribute()" is deprecated, return an instance of "%s" instead.', static::class, Vote::class);
```
```suggestion
                trigger_deprecation('symfony/security-core', '5.4', 'Returning a boolean in "%s::voteOnAttribute()" is deprecated, return an instance of "%s" instead.', static::class, Vote::class);
```

noniagriconomie replied:
in other files as well

yellow1912 replied:
Sorry, too many files. I fixed all that I can find now. Please let me know if I miss anything.

noniagriconomie commented on src/Symfony/Component/Security/Core/Authorization/Voter/Vote.php:
```diff
@@ -0,0 +1,91 @@
+    /** @var int One of the VoterInterface::ACCESS_* constants */
+    private $access;
+    private $reason;
+    private $parameters;
```
move like `$context` than `$parameters` no?

yellow1912 replied:
sure, I will change this.

noniagriconomie commented on src/Symfony/Component/Security/Core/Authorization/Voter/Vote.php:
```diff
@@ -0,0 +1,91 @@
+{
+    /** @var int One of the VoterInterface::ACCESS_* constants */
+    private $access;
+    private $reason;
```
maybe we should call it `$message` even if it is the `the ability for voter to return decision reason` ?
as we use it to populate the exception message later

yellow1912 replied:
sure, I don't have any problem with that. There will be a few places I have to change as well if you think this change is necessary. Personally I think reason makes sense, but I can see why you want to use message.

noniagriconomie commented on src/Symfony/Component/Security/Core/Event/VoteEvent.php:
```diff
@@ -28,11 +29,20 @@ final class VoteEvent extends Event
+    /**
+     * @param Vote|int $vote
+     */
+    public function __construct(VoterInterface $voter, $subject, array $attributes, $vote)
```
is it allowed to do so regarding BC and code evolution?
what about creating a new event class, and mark this one as deprecated in 5.4?
like it was done for kernel events in the past iirc

yellow1912 replied:
I honestly don't know. The code is copied "as is" from the other PR (since the original contributor is busy now). I'm willing to make any changes to comply with the standards. Please let me know your recommendation.

noniagriconomie replied:
yes I know you are continuing the work and thank you for that :)
perhaps for this point we should wait a core member to give us the `reason` 🥁

yellow1912 replied:
@stof Can you help to give some recommendations on how to handle this? Thank you very much.

chalasr replied:
BC-wise, this specific change is fine: it's a final constructor so there is no potential signature mismatch, and it keeps handling both types.

noniagriconomie commented on src/Symfony/Component/Security/Core/Authorization/AccessDecisionManagerInterface.php:
```diff
@@ -17,16 +17,18 @@
+     * @param array  $attributes An array of attributes associated with the method being invoked
+     * @param object $object     The object to secure
      *
-     * @return bool
```
maybe keep this comment?

yellow1912 committed:
Fix deprecation version

yellow1912 committed:
Rename vote $reason to $message, $parameters to $context

yellow1912 committed:
Add back return typehint in phpdoc

yellow1912 committed:
Fix coding standard

yellow1912 committed:
Add more checkings for deprecation notices inside tests

yellow1912 committed:
Merge test code with latest 5.4

chalasr commented on src/Symfony/Component/Security/Core/Exception/AccessDeniedException.php:
```diff
@@ -57,4 +62,26 @@ public function setSubject($subject)
+        $this->message .= ' '.rtrim(' '.implode(' ', $messages));
+    }
+
+    public function getAccessDecision(): AccessDecision
```
This will error if `setAccessDecision` was not called
```suggestion
    public function getAccessDecision(): ?AccessDecision
```

the property docblock needs to be updated accordingly

yellow1912 committed:
Allow getAccessDecision to return null in AccessDeniedException

yellow1912 commented:
@noniagriconomie can you check and let me know if there is anything else I should fix? I see unittests still fail but not entirely sure why, I see only notices.

yellow1912 committed:
Merge security.html.twig with new changes

noniagriconomie commented:
@yellow1912 unfortunatly, i may not be the best to answer `if there is anything else I should fix` here
maybe ping some core maintainers and/or code owners :)
but iirc the v5.4 is feature freezed

yellow1912 committed:
Fix coding standards

yellow1912 committed:
Add missing dot to comply with standards

yellow1912 commented:
@stof Please let me know if:

1. There is anything else I should fix?
2. Should I bump this to another branch if 5.4 is frozen? And which one should that be?

chalasr commented:
I'll take time to test this thoroughly during the last week of October. My main concern is about the BC layer, for 3rd party bundles especially, but I feel like we are quite close.

yellow1912 commented:
@chalasr Have you got the time to take a look at this? Once you checked everything I can update it to 6.x instead because it seems like 5.x is closed now.

chalasr commented:
@yellow1912 We will need to make this happen on 6.1, yes.

The core team' focus is currently about stabilizing core code/docs and syncing community projects for 5.4 and 6.0.

I will revisit this early in the 6.1 development stage. Thanks for your patience.

derrabus commented:
In the meantime, you can already rebase onto the 6.0 branch (6.1 is not open yet) which also means you can use all the juicy new PHP 8.0 features. 😎

noniagriconomie commented:
@yellow1912 I've reminded myself this PR, did you had time/energy to rebase it for sf v6?

thank you :)

yellow1912 commented:
> @yellow1912 I've reminded myself this PR, did you had time/energy to rebase it for sf v6?
> 
> thank you :)

I haven't got the time to do that, but I will try to finish it within the next 2 weeks, too many things on my disk at the moment.

fabpot commented:
Closing in favor of #46493
